GCInfo: Support versioning.

swaroop-sridhar opened the pull request:
This change enables the VM to support multiple versions GCInfo concurrently.
This is necessary in light of upcoming work to add ReturnType and other
modifications to the GCInfo format -- so that existing ReadyToRun images
will continue to run correctly.

The version# is not stored in the GcInfo structure -- because it is
wasteful to store the version once for every method. Instead, it is
tracked per range-section of generated/loaded methods.

The GCInfo version is computed as:
1) The current GCINFO_VERSION for JITted and Ngened images
2) A function of the Ready-to-run major version stored in READYTORUN_HEADER
   for ready-to-run images. ReadyToRunJitManager::JitTokenToGCInfoVersion()
   provides the GcInfo version for any Method. Currently, there's only one
   version of GCInfo.

An abstraction GCTable is added to the GcInfo interface, which tracks the
{GcInfoBlock, GcInfoVersion} pair in-memory. Several GcInfo APIs are
modified to use GcTable in place of GcInfo pointers.

Notes:
1) SOS GcDump: The GCDump API has separate dump routines for Header and the
pointer-liveness information (DumpGCTable and DumpGCHeader) each of which
advance a pointer to the GCInfo block. These APIs are not changed to
recieve a GCTable in place of the GcInfo block pointer. Instead, they
recieve the GcInfo version at the time of construction.
2) Some routines that are specific to x86 GcTables (ex: crackMethodInfoHdr)
are not yet updated to use versioning, since the development plan is to
update the Non-x86 GcInfo structure first.

swaroop-sridhar commented:
@dotnet-bot  test Ubuntu x64 Checked Build and Test

swaroop-sridhar commented:
@jkotas please review

swaroop-sridhar commented:
cc: @dotnet/jit-contrib

jkotas commented:
We have always called this GCInfo. What's the point in renaming it to GCTable?

swaroop-sridhar commented:
GCInfo is the data format written to the disk -- that structure is unchanged still called GcInfo.
GCTable is only the in-memory container that holds {GcInfo pointer, and Version#}.
So, its not a renaming. We can call the wrapper struct something else, if its confusing.

jkotas commented:
Using `GCTable` name for `{GcInfo pointer, and Version#}` sounds confusing to me - I do not see the table part. I think the name should still have `GCInfo` in it. Would it make sense to call it just `GCInfo`? Or something like `GCInfoBlob`?

swaroop-sridhar commented:
@jkotas: I renamed the `GCTable` structure as `GCInfoToken` and updated all the APIs appropriately. I did not call it simply `GCInfo`, because there's another struct called `GCInfo` defined by the jIT.

jkotas commented on src/vm/util.cpp:
```diff
@@ -3130,13 +3130,13 @@ BOOL JITNotifications::UpdateOutOfProcTable()
 
 GPTR_IMPL(GcNotification, g_pGcNotificationTable);
 
-GcNotifications::GcNotifications(GcNotification *gcTable)
```
This looks like find&replace mistake. This class has nothing to with GCInfo

jkotas commented:
LGTM otherwise.
